### ISSUE TITLE
loki gateway in loki-simple-scalable not able to be added to grafana as loki datasource

### ISSUE THREAD

**Issue by craustin** [grafana/helm-charts#921]
I'm using loki-simple-scalable chart v0.1.0 to install loki on Rancher.
When I create a Loki datasource in Grafana v7.5.11 pointing at the gateway's ingress hostname, Grafana reports:
> Loki: Cannot connect to Loki. Unexpected error
In the gateway's log (nginx log) it reports an HTTP 405 response to a request for `/loki/api/v1/label?start=1640728988233000000`

If I instead add a Loki datasource for the `simple-scalable-read` service's internal dns hostname instead (on port 3100), Grafana adds the datasource successfully.

I tried modifying the gateway's nginx conf to return a 200 for an OPTIONS request on this route, but that didn't seem to help - and confusingly, it seems the read service returns the same 405 for that OPTIONS request.

Is it expected that Grafana would fail to add a Loki datasource using the gateway hostname?

**Comment by rufreakde** [comment]
I would like to Bumb this issue I tried to use:
```http://loki-gateway.up-monitoring.svc.cluster.local:3100``` which is mentioned in the official guide:
https://grafana.com/docs/loki/latest/installation/simple-scalable-helm/

But it fails when I press save and test
<img width="777" alt="Screenshot 2022-06-27 at 17 51 50" src="https://user-images.githubusercontent.com/28104368/175981607-a8da02a1-4db2-4e8b-837f-223d83b30e5b.png">
!

**Comment by rufreakde** [comment]
changed to helm chart distributed that worked...
```http://monitoring-loki-loki-distributed-gateway.monitoring.svc.cluster.local/```

**Comment by trevorwhitney** [comment]
I believe the `<helm-installation-name>` has been removed from service name. Does it work if you try `loki-gateway.<namespace>.svc.cluster.local` (or `enterprise-logs-gateway.<namespace>.svc.cluster.local` when `enterprise.enabled: true`)?

**Comment by rpf3** [comment]
I think I was running into the same issue as everyone here.  I am running the standalone Grafana and Loki "simple scalable" charts but could not successfully add Loki as a data source.  What I ended up tracking down was there is a chart value `loki.auth_enabled` that defaults to `true`.  This default value configures Loki in multi-tenant mode which requires all requests to be sent with an auth header.  I came to this conclusion by tailing the logs of the "loki-gateway" service and saw a bunch of 401 errors.  Once I deployed with `loki.auth_enabled` set to `false` everything worked as expected.

refs:
* https://grafana.com/docs/loki/latest/operations/authentication/
* https://grafana.com/docs/loki/latest/operations/multi-tenancy/
* https://grafana.com/docs/loki/latest/configuration/#supported-contents-and-default-values-of-lokiyaml

**Comment by rufreakde** [comment]
> loki-gateway.<namespace>.svc.cluster.local

loki-gateway.<namespace>.svc.cluster.local this worked in the new chart as well.

@trevorwhitney can this be set as default value in the chart?

**Comment by trevorwhitney** [comment]
that seems reasonable, would you mind making a PR? would be better to discuss in context I think with the code to see how it affects different code paths.

**Comment by rpf3** [comment]
@trevorwhitney do you mean a PR making the `loki.auth_enabled` default to `false`?

**Comment by trevorwhitney** [comment]
no, that was in response to the following (I think, as this was over a month ago)

> loki-gateway..svc.cluster.local this worked in the new chart as well.

> @trevorwhitney can this be set as default value in the chart?

**Comment by rufreakde** [comment]
@trevorwhitney sorry somehow I missed this thread. I will set myself a reminder to check that chart.

Modifying the values.yaml should not be that hard. I switched personally away from „simple-scalable“ helm chart and use just „loki“ now. With my own S3 db.

**Comment by trevorwhitney** [comment]
@rufreakde glad to hear that, we're hoping to get everyone on to the new `loki` chart in the `grafana/loki` repo so everyone's on the same chart and we can maximize benefit of community contributions (rather than having them split between multiple charts)

**Comment by hpl002** [comment]
TLDR for the next fella;

1. Disable auth, as discovered
 https://github.com/grafana/helm-charts/issues/921#issuecomment-1183597102
2. Properly format URL -> http://loki-gateway.<NAMESPACE>.svc.cluster.local
https://github.com/grafana/helm-charts/issues/921#issuecomment-1285575638

**Comment by raxtorDD** [comment]
> @trevorwhitney sorry somehow I missed this thread. I will set myself a reminder to check that chart.
> 
> Modifying the values.yaml should not be that hard. I switched personally away from „simple-scalable“ helm chart and use just „loki“ now. With my own S3 db.

I am trying to use the "normal" loki chart and can't figure out why Grafana can't connect to this source.

I tried to add following url in Grafana: "http://loki-gateway.loki-dev.svc.cluster.local" also with auth_enabled false.
Grafana has been deployed to namespace "grafana-dev".

Could you share your value files? Thanks!

**Comment by rufreakde** [comment]
> @rufreakde
> 
> I am trying to use the "normal" loki chart and can't figure out why Grafana can't connect to this source.
> 
> I tried to add following url in Grafana: "http://loki-gateway.loki-dev.svc.cluster.local" also with auth_enabled false. Grafana has been deployed to namespace "grafana-dev".
> 
> Could you share your value files? Thanks!

Usually this would not be an issue but unfortunately I am on vacation now. Quickest would be some time around january.

**Comment by raxtorDD** [comment]
> > @rufreakde
> > I am trying to use the "normal" loki chart and can't figure out why Grafana can't connect to this source.
> > I tried to add following url in Grafana: "http://loki-gateway.loki-dev.svc.cluster.local" also with auth_enabled false. Grafana has been deployed to namespace "grafana-dev".
> > Could you share your value files? Thanks!
> 
> Usually this would not be an issue but unfortunately I am on vacation now. Quickest would be some time around january.

No worries! Maybe I have found a solution until then.
What really grinds my gears is that the gateway default deployment is set to port 80. Reader, Writer do have port 3100. Shouldn't the gateway have port 3100?

**Comment by raxtorDD** [comment]
So I got it working.
Turns out that only decreasing read and write replicas isn't enough.
For my dev instance I only needed one replica of read and write and not this high scalable option with 3 replicas, each on a different node. I had to add "replication_factor" and set it to "1" too.

I am adding my value files, if others experiencing the same issue.
Grafana, Loki and Promtail are deployed into the same namespace "monitoring-dev".
This is all running on an azure kubernetes cluster with Kong Gateway deployed.
Loki uses azure blob storage.

Hope this helps somebody.

```
# loki values-dev.yaml

loki:
  auth_enabled: false
  server:
    http_listen_port: 3100
  commonConfig:
    replication_factor: 1
write:
  replicas: 1
read:
  replicas: 1

schema_config:
  configs:
  - from: "2020-12-11"
    index:
      period: 24h
      prefix: loki_index_
    schema: v11
    store: boltdb-shipper
    object_store: azure
storage_config:
  azure:
    use_service_principal: true
    tenant_id : <tenant_id>
    client_id: <client_id>
    client_secret: <client_secret>
    container_name: loki-dev
    request_timeout: 0
  boltdb_shipper:
    active_index_directory: /var/loki/index
    cache_location: /var/loki/cache
    cache_ttl: 24h
    shared_store: azure
```

```
# Grafana values-dev.yaml

service:
  enabled: true
  type: ClusterIP

ingress:
  enabled: true
  annotations:
    kubernetes.io/ingress.class: kong
  path: /monitoring/grafana
  hosts:
    - <hosturl>
  tls:
    - secretName: acme-http01-cert-issued-default
    - hosts:
        - <hosturl>

grafana.ini:
  server:
    root_url: <url>
    serve_from_sub_path: true
  auth.azuread:
    name: Azure AD
    enabled: true
    allow_sign_up: true
    client_id: <clientid>
    client_secret: <clientsecret>
    scopes: openid email profile offline_access
    auth_url: https://login.microsoftonline.com/<tenantid>/oauth2/v2.0/authorize
    token_url: https://login.microsoftonline.com/<tenantid>/oauth2/v2.0/token
    allowed_groups: <groupid>
    role_attribute_strict: true
    allow_assign_grafana_admin: true

datasources:
  datasources.yaml:
    apiVersion: 1
    datasources:
    - name: Loki
      type: loki
      url: http://loki-gateway.monitoring-dev.svc.cluster.local
      access: proxy
      isDefault: true
```

```
# Promtail values-dev.yaml

config:
  clients:
    - url: http://loki-gateway/loki/api/v1/push
```

**Comment by david-nano** [comment]
Still issue in here.
Here is my `values.yaml`:
```
loki:
  enabled: true
  auth_enabled: false
  global:
    image:
      registry: localregsitrey.local:8084
  ingress:
    enabled: true
    ingressClassName: "nginx"
    paths:
      write:
        - /api/prom/push
        - /loki/api/v1/push
      read:
        - /api/prom/tail
        - /loki/api/v1/tail
        - /loki/api
        - /api/prom/rules
        - /loki/api/v1/rules
        - /prometheus/api/v1/rules
        - /prometheus/api/v1/alerts
      singleBinary:
        - /api/prom/push
        - /loki/api/v1/push
        - /api/prom/tail
        - /loki/api/v1/tail
        - /loki/api
        - /api/prom/rules
        - /loki/api/v1/rules
        - /prometheus/api/v1/rules
        - /prometheus/api/v1/alerts
    hosts:
      - loki.server.com
  write:
    persistence:
      storageClass: "netapp-storage"
  read:
    persistence:
      storageClass: "netapp-storage"
  backend:
    persistence:
      storageClass: "netapp-storage"
  singleBinary:
    persistence:
      storageClass: "netapp-storage"
  commonConfig:
    replication_factor: 1
  storage:
    type: 'filesystem'

singleBinary:
  replicas: 1
```

Here is my services:
```
$ kubectl -n loki get svc
NAME                        TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
loki-backend                ClusterIP   10.43.4.250     <none>        3100/TCP,9095/TCP   3h21m
loki-backend-headless       ClusterIP   None            <none>        3100/TCP,9095/TCP   3h21m
loki-canary                 ClusterIP   10.43.88.231    <none>        3500/TCP            3h21m
loki-gateway                ClusterIP   10.43.95.105    <none>        80/TCP              3h21m
loki-memberlist             ClusterIP   None            <none>        7946/TCP            3h21m
loki-read                   ClusterIP   10.43.212.172   <none>        3100/TCP,9095/TCP   3h21m
loki-read-headless          ClusterIP   None            <none>        3100/TCP,9095/TCP   3h21m
loki-write                  ClusterIP   10.43.14.241    <none>        3100/TCP,9095/TCP   3h21m
loki-write-headless         ClusterIP   None            <none>        3100/TCP,9095/TCP   3h21m
query-scheduler-discovery   ClusterIP   None            <none>        3100/TCP,9095/TCP   3h21m
```


I tried to add it into Grafana data source: `http://loki-gateway.loki.svc.cluster.local` after reading: https://grafana.com/docs/loki/latest/installation/helm/concepts/#gateway
and yes, `auth_enabled=false` as you can see.
My error is the same:
`Unable to fetch labels from Loki (Failed to call resource), please check the server logs for more details`

Anyone figure this out? Do we need to use a gateway? do we need to provide a port? I see the gateway uses default 80 to manage all traffic.

**Comment by romosa** [comment]
I am still not able to add loki as datasource.

I already tried suggested configurations here.

Loki url: `http://loki-backend.monitoring-system.svc.cluster.local:3100/`
Here is values.yaml
```
write:
  replicas: 1
read:
  replicas: 1

loki:
  auth_enabled: false
  server:
    http_listen_port: 3100
  commonConfig:
    replication_factor: 1
  limits_config:
    reject_old_samples: true
    reject_old_samples_max_age: 168h
    max_cache_freshness_per_query: 30m
    split_queries_by_interval: 0
```

I see this logs on gateway.
```
level=error ts=2024-03-26T03:47:51.957257389Z caller=cached_client.go:189 msg="failed to build table names cache" err="RequestCanceled: request context canceled\ncaused by: context canceled"
level=error ts=2024-03-26T03:47:51.95729556Z caller=compactor.go:523 msg="failed to run compaction" err="failed to list tables: RequestCanceled: request context canceled\ncaused by: context canceled"
2024/03/26 03:48:13 WARN: failed to get session token, falling back to IMDSv1: 405 Method Not Allowed: Method Not Allowed
        status code: 405, request id:
caused by: EC2MetadataError: failed to make EC2Metadata request
<html>
<head><title>405 Not Allowed</title></head>
<body bgcolor="white">
<center><h1>405 Not Allowed</h1></center>
</body>
</html>

        status code: 405, request id:
level=error ts=2024-03-26T03:48:13.672687062Z caller=reporter.go:205 msg="failed to delete corrupted cluster seed file, deleting it" err="NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors"
level=error ts=2024-03-26T03:48:42.012564069Z caller=ruler.go:571 msg="unable to list rules" err="NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors"

```
